### DISCUSSION TITLE
Support GitHub custom properties for autodiscoverFilter

### DISCUSSION THREAD

**Discussion by beatngu13** [renovatebot/renovate#31599] [Suggest an Idea]
### Tell us more.

Via [GitHub's docs](https://docs.github.com/en/organizations/managing-organization-settings/managing-custom-properties-for-repositories-in-your-organization):

>Custom properties allow you to decorate your repositories with information such as compliance frameworks, data sensitivity, or project details. Custom properties are private and can only be viewed by people with read permissions to the repository.

That is, properties are similar to [topics](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/classifying-your-repository-with-topics), although they serve different purposes.

Today, Renovate already offers [`autodiscoverTopics`](https://docs.renovatebot.com/self-hosted-configuration/#autodiscovertopics) to select repositories to be updated. It would be super helpful if Renovate would also support GitHub custom properties with something like `autodiscoverCustomProperties`.

**Comment by rarkins** [comment]
Would it be essentially the same functionality as autodiscoverTopics except that it's querying a different list from the github API? And are these properties returned in the same API which Renovate already uses and gets topics from?

**Reply by beatngu13** [reply]
>Would it be essentially the same functionality as autodiscoverTopics except that it's querying a different list from the github API?

I think so. How `autodiscoverTopics` works today:

```json
{
  "autodiscoverTopics": [
    "some-topic",
    "another-topic"
  ]
}
```

Then Renovate will update all repositories that have the specified topics (I believe it's logical `OR`?). I guess for `autodiscoverCustomProperties` it should behave the same:

```json
{
  "autodiscoverCustomProperties": [
    "some-property=some-value",
    "another-property=another-value"
  ]
}
```

Or:

```json
{
  "autodiscoverCustomProperties": [
    { "some-property": "some-value" },
    { "another-property": "another-value" }
  ]
}
```

Or any other syntax you think is appropriate.

>And are these properties returned in the same API which Renovate already uses and gets topics from?

Both are under `/repos/OWNER/REPO`, but the endpoints are different. This is how [topics can be queried](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#get-all-repository-topics):

```bash
$ curl -L \
  -H "Accept: application/vnd.github+json" \
  -H "Authorization: Bearer <YOUR-TOKEN>" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  https://api.github.com/repos/OWNER/REPO/topics
```

Response:

```json
{
  "names": [
    "octocat",
    "atom",
    "electron",
    "api"
  ]
}
```

And this is how [custom properties can be queried](https://docs.github.com/en/rest/repos/custom-properties?apiVersion=2022-11-28#get-all-custom-property-values-for-a-repository):

```bash
curl -L \
  -H "Accept: application/vnd.github+json" \
  -H "Authorization: Bearer <YOUR-TOKEN>" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  https://api.github.com/repos/OWNER/REPO/properties/values
```

Response:

```json
[
  {
    "property_name": "environment",
    "value": "production"
  },
  {
    "property_name": "service",
    "value": "web"
  },
  {
    "property_name": "team",
    "value": "octocat"
  }
]
```

**Reply by rarkins** [reply]
That's not how topics filtering is done today.

See https://github.com/renovatebot/renovate/blob/161ebccd2d17c6f6b05ff3f7da05c27def995ebf/lib/modules/platform/github/index.ts#L296-L298 as a starting point. Topics are returned from the user/repos endpoint, which means they don't need to be queried once per repo. The endpoint is documented here: https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#list-repositories-for-the-authenticated-user

It doesn't looks like custom properties are returned by that endpoint, which would mean that Renovate needs to do one API call per repo to do additional filtering. I would expect that custom properties are more likely to be used by large organizations, which also means it would use up a lot of API calls to perform the filtering.

**Reply by beatngu13** [reply]
I see. Doing one additional API call per repository is … meh.

I will open a discussion over at [GitHub community](https://github.com/orgs/community/discussions) / raise a support ticket.

**EDIT:** https://github.com/orgs/community/discussions/139846
